### ISSUE TITLE
Feature Request: better cross entropy loss CUDA kernel

### ISSUE THREAD

**Issue by JohannesGaessler** [ggerganov/llama.cpp#10467]
### Prerequisites

- [X] I am running the latest code. Mention the version if possible as well.
- [X] I carefully followed the [README.md](https://github.com/ggerganov/llama.cpp/blob/master/README.md).
- [X] I searched using keywords relevant to my issue to make sure that I am creating a new issue that is not already open (or closed).
- [X] I reviewed the [Discussions](https://github.com/ggerganov/llama.cpp/discussions), and have a new and useful enhancement to share.

### Feature Description

The CUDA cross entropy loss kernel needs to be re-written. I am currently working on this and making an issue to avoid duplicating work with @slaren .

### Motivation

The current cross entropy loss kernel needs a lot of shared memory and thus can't handle input shapes with a large number of classes. This is a problem because LLMs with their large vocabularies effectively have a large number of classes. Currently, due to calling `DeviceReduce::Sum` afterwards the calculation just silently fails for those cases.

### Possible Implementation

I will do an inefficient but easy implementation. Long-term I think a FlashAttention-like implementation can be done that calculates the cross entropy loss in a single pass over the logits/labels at the cost of some rescaling.
